File: backend/services/deepseekService.js
import axios from 'axios';
import FormData from 'form-data';
import fs from 'fs';
import dotenv from 'dotenv';
import { DEEPSEEK_API_KEY } from '../config/config.js';

// Load environment variables
dotenv.config();

class DeepseekService {
  constructor() {
    // DeepSeek API v3配置
    // API文档: https://api-docs.deepseek.com/zh-cn/
    this.apiKey = DEEPSEEK_API_KEY;
    if (!this.apiKey) {
      console.warn('[DeepSeek] API key not found in environment variables');
      throw new Error('DeepSeek API key is required');
    }
    this.apiUrl = 'https://api.deepseek.com/v3';
    this.client = axios.create({
      baseURL: this.apiUrl,
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
        'Accept': 'application/json',
        'Content-Type': 'application/json'
      },
      validateStatus: function (status) {
        return status >= 200 && status < 500;
      },
      timeout: 30000 // 30秒超时
    });

    // 添加请求拦截器用于日志记录
    this.client.interceptors.request.use(config => {
      console.log(`[DeepSeek API Request] ${config.method.toUpperCase()} ${config.url}`);
      return config;
    });

    // 添加响应拦截器用于错误处理
    this.client.interceptors.response.use(
      response => {
        console.log(`[DeepSeek API Response] Status: ${response.status}`);
        return response;
      },
      error => {
        console.error('[DeepSeek API Error]', error.response?.data || error.message);
        throw error;
      }
    );
  }

  // 分析提示词模板
  static ANALYSIS_PROMPTS = {
    // 图片分析提示词
    general: `请分析这张图片中的健康相关信息，包括但不限于：
1. 主要健康指标
2. 可能存在的健康风险
3. 建议采取的措施
请用专业但易懂的语言描述，并给出具体的改善建议。`,

    medical_report: `请分析这份医疗报告图片，重点关注：
1. 关键检测指标及其含义
2. 异常值分析
3. 健康风险评估
4. 建议的后续措施
请用专业的医学术语解释，同时提供通俗的解释。`,

    lifestyle: `请分析这张生活方式相关的图片，关注：
1. 当前的生活习惯对健康的影响
2. 潜在的健康风险
3. 改善建议
4. 可以采取的具体行动
请提供实用的、可执行的建议。`,

    // 文本分析提示词
    text_health: `作为专业的健康顾问，请对以下健康相关文本进行全面分析，重点关注：

1. 健康状况综述
   - 整体健康评估
   - 关键健康指标分析
   - 生活方式评估

2. 风险分析
   - 当前存在的健康风险
   - 潜在的健康隐患
   - 风险等级评估（低/中/高）

3. 改善建议
   - 生活方式调整建议
   - 饮食营养建议
   - 运动建议
   - 作息时间建议

4. 后续跟进
   - 需要定期监测的指标
   - 建议进行的健康检查
   - 预防保健措施

请以专业但通俗易懂的语言提供分析结果，确保建议具体且可执行。`,

    text_medical: `作为专业的医疗顾问，请对以下医疗检验报告进行专业解读，重点关注：

1. 检验指标解读
   - 异常指标识别和分析
   - 指标间关联性分析
   - 参考值范围对比

2. 健康风险评估
   - 当前健康问题诊断
   - 潜在疾病风险评估
   - 并发症风险分析

3. 诊疗建议
   - 建议就医科室
   - 建议进行的进一步检查
   - 用药建议（如适用）
   - 生活方式调整建议

4. 预防保健措施
   - 短期改善建议
   - 长期健康管理计划
   - 定期复查建议

请使用专业但易于理解的语言，提供详细的分析和明确的建议。对于异常指标，请特别说明其可能的原因和影响。`
  };

  /**
   * 分析图片
   * @param {string} imagePath - 图片文件路径
   * @param {string} analysisType - 分析类型（general/medical_report/lifestyle）
   * @returns {Promise<Object>} 分析结果
   */
  async analyzeImage(imagePath, analysisType = 'general') {
    try {
      // 读取图片文件
      const imageBuffer = fs.readFileSync(imagePath);
      const formData = new FormData();
      
      // 添加图片文件
      formData.append('image', imageBuffer, {
        filename: 'image.jpg',
        contentType: 'image/jpeg'
      });

      // 添加提示词
      formData.append('prompt', DeepseekService.ANALYSIS_PROMPTS[analysisType]);

      // 调用 DeepSeek API
      const response = await this.client.post('/vision/completions', formData, {
        headers: {
          ...formData.getHeaders()
        }
      });

      // 处理响应
      return this.processAnalysisResponse(response.data);
    } catch (error) {
      console.error('图片分析失败:', error);
      throw new Error(error.response?.data?.error || '图片分析失败');
    }
  }

  /**
   * 处理分析响应
   * @param {Object} rawResponse - API 原始响应
   * @returns {Object} 处理后的分析结果
   */
  processAnalysisResponse(rawResponse) {
    // 提取关键信息
    const {
      analysis,
      confidence,
      recommendations,
      risks
    } = rawResponse;

    // 结构化返回结果
    return {
      summary: analysis.summary,
      confidence: confidence,
      metrics: {
        healthScore: this.calculateHealthScore(analysis),
        riskLevel: this.assessRiskLevel(risks),
        reliabilityScore: confidence
      },
      recommendations: recommendations.map(rec => ({
        category: rec.category,
        suggestion: rec.text,
        priority: rec.priority
      })),
      risks: risks.map(risk => ({
        type: risk.type,
        severity: risk.severity,
        description: risk.description
      })),
      rawAnalysis: analysis
    };
  }

  /**
   * 计算健康评分
   * @param {Object} analysis - 分析结果
   * @returns {number} 健康评分（0-100）
   */
  calculateHealthScore(analysis) {
    // 基于多个因素计算综合健康评分
    const factors = {
      generalHealth: analysis.generalHealthScore || 70,
      riskFactors: analysis.riskFactorsScore || 80,
      lifestyle: analysis.lifestyleScore || 75
    };

    const weights = {
      generalHealth: 0.4,
      riskFactors: 0.35,
      lifestyle: 0.25
    };

    return Object.keys(factors).reduce((score, factor) => {
      return score + (factors[factor] * weights[factor]);
    }, 0);
  }

  /**
   * 评估风险等级
   * @param {Array} risks - 风险列表
   * @returns {string} 风险等级（low/medium/high）
   */
  assessRiskLevel(risks) {
    const severityScores = {
      low: 1,
      medium: 2,
      high: 3
    };

    const averageSeverity = risks.reduce((sum, risk) => {
      return sum + severityScores[risk.severity.toLowerCase()];
    }, 0) / risks.length;

    if (averageSeverity <= 1.5) return 'low';
    if (averageSeverity <= 2.5) return 'medium';
    return 'high';
  }

  /**
   * 分析文本内容
   * @param {string} text - 要分析的文本内容
   * @param {string} analysisType - 分析类型（text_health/text_medical）
   * @returns {Promise<Object>} 分析结果
   */
  async analyzeText(text, analysisType = 'text_health') {
    try {
      // 调用 DeepSeek API v3
      const response = await this.client.post('/chat/completions', {
        model: 'deepseek-chat',
        messages: [
          {
            role: 'system',
            content: '你是一个专业的健康顾问，负责分析健康数据并提供专业的建议。请以JSON格式返回分析结果。'
          },
          {
            role: 'user',
            content: `${DeepseekService.ANALYSIS_PROMPTS[analysisType]}\n\n${text}`
          }
        ],
        max_tokens: 2000,
        temperature: 0.3, // 降低随机性以获得更稳定的结果
        top_p: 0.8,
        frequency_penalty: 0.1,
        presence_penalty: 0.1,
        response_format: { type: 'json_object' }
      });

      // 处理响应
      const responseData = response.data;
      if (!responseData.choices?.[0]?.message?.content) {
        return {
          summary: '无法解析响应',
          recommendations: [],
          risks: [],
          metrics: {
            healthScore: 75,
            stressLevel: 'medium',
            sleepQuality: 'medium'
          }
        };
      }
      return this.processTextResponse(responseData);
    } catch (error) {
      console.error('文本分析失败:', error);
      if (error.message === 'Invalid API response format') {
        throw error;
      }
      throw new Error(error.response?.data?.error || '文本分析失败');
    }
  }

  /**
   * 处理文本分析响应
   * @param {Object} rawResponse - API 原始响应
   * @returns {Object} 处理后的分析结果
   */
  processTextResponse(rawResponse) {
    try {
      const content = JSON.parse(rawResponse.choices[0].message.content);
      return {
        summary: content.summary,
        recommendations: content.recommendations.map(rec => typeof rec === 'string' ? rec : rec.text),
        riskFactors: content.risks.map(risk => typeof risk === 'string' ? risk : risk.description),
        metrics: {
          healthScore: content.generalHealthScore || 85,
          stressLevel: content.metrics?.stressLevel || 'medium',
          sleepQuality: content.metrics?.sleepQuality || 'medium'
        }
      };
    } catch (error) {
      console.error('[DeepSeek] 处理响应失败:', error);
      // 返回用户友好的错误信息
      throw new Error(`处理健康数据失败: ${error.message}`);
    }
  }

  /**
   * 从文本响应中提取结构化信息
   * @param {string} text - DeepSeek响应文本
   * @returns {Object} 结构化信息
   */
  extractStructuredInfo(text) {
    // 初始化默认结构
    const defaultStructure = {
      summary: '',
      recommendations: [],
      risks: [],
      generalHealthScore: 75,
      riskFactorsScore: 75,
      lifestyleScore: 75
    };

    try {
      // 尝试将文本解析为JSON（如果DeepSeek返回JSON格式）
      const parsed = JSON.parse(text);
      return {
        ...defaultStructure,
        ...parsed
      };
    } catch (e) {
      // 如果不是JSON，进行文本分析提取结构化信息
      const sections = text.split('\n\n');
      const structure = { ...defaultStructure };

      // 提取摘要（第一段）
      structure.summary = sections[0] || '';

      // 提取建议和风险（根据关键词）
      sections.forEach(section => {
        if (section.toLowerCase().includes('建议') || section.toLowerCase().includes('推荐')) {
          structure.recommendations.push({
            text: section,
            category: '一般建议',
            priority: 'medium'
          });
        } else if (section.toLowerCase().includes('风险') || section.toLowerCase().includes('警告')) {
          structure.risks.push({
            description: section,
            type: '健康风险',
            severity: 'medium'
          });
        }
      });

      return structure;
    }
  }
}

export default new DeepseekService();

File: dna-backend/app/services/deepseek_service.py
import httpx
import logging
from fastapi import HTTPException
import os
from dotenv import load_dotenv
import json
from .mock_deepseek_service import mock_analyze_sequence, MOCK_RESPONSES

load_dotenv()

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

DEEPSEEK_API_KEY = os.getenv("DEEPSEEK_API_KEY")
if not DEEPSEEK_API_KEY:
    logger.error("DeepSeek API key not found in environment variables")
    raise ValueError("DeepSeek API key not configured. Please set DEEPSEEK_API_KEY environment variable.")

CLAUDE_API_KEY = os.getenv("CLAUDE_API_KEY")

async def analyze_sequence(
    sequence: str,
    provider: str = "deepseek",
    analysis_type: str = "health",
    include_recommendations: bool = True,
    include_risk_factors: bool = True,
    include_metrics: bool = True
) -> dict:
    if os.getenv("MOCK_DEEPSEEK_API"):
        if not sequence:
            raise HTTPException(status_code=400, detail="No sequence provided")
        return await mock_analyze_sequence(sequence, analysis_type)
    try:
        logger.info(f"Starting analysis with provider: {provider}")
        logger.info(f"Sequence length: {len(sequence)}")
        
        if not sequence:
            logger.error("Empty sequence provided")
            raise HTTPException(
                status_code=400,
                detail="Sequence cannot be empty"
            )
        
        if provider == "claude":
            logger.info("Using Claude provider")
            return await analyze_with_claude(sequence)
        elif provider == "deepseek":
            logger.info("Using DeepSeek provider")
            return await analyze_with_deepseek(sequence)
        else:
            logger.error(f"Unsupported provider: {provider}")
            raise HTTPException(
                status_code=400,
                detail="Invalid provider specified"
            )
    except HTTPException as e:
        logger.error(f"HTTP Exception in analyze_sequence: {str(e)}")
        raise
    except Exception as e:
        logger.error(f"Unexpected error in analyze_sequence: {str(e)}")
        return {
            "success": False,
            "error": f"Failed to analyze sequence: {str(e)}"
        }

async def analyze_with_claude(sequence: str) -> dict:
    if not CLAUDE_API_KEY:
        raise HTTPException(
            status_code=500,
            detail="Claude API key is not configured"
        )
    
    headers = {
        "x-api-key": CLAUDE_API_KEY,
        "Content-Type": "application/json"
    }
    
    data = {
        "model": "claude-3-opus-20240229",
        "messages": [
            {
                "role": "system",
                "content": "You are a bioinformatics expert analyzing biological sequences. Provide detailed analysis including sequence type identification, features, health implications, and recommendations."
            },
            {
                "role": "user", 
                "content": f"Please analyze this sequence and provide health insights: {sequence}"
            }
        ],
        "temperature": 0.3,
        "max_tokens": 2000
    }

    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.post(
                "https://api.anthropic.com/v1/messages",
                json=data,
                headers=headers
            )
            
            if response.status_code == 200:
                result = response.json()
                return {
                    "success": True,
                    "analysis": {
                        "summary": result["content"][0]["text"],
                        "provider": "claude",
                        "model": "claude-3-opus-20240229"
                    }
                }
            else:
                raise HTTPException(
                    status_code=response.status_code,
                    detail=f"Claude API error: {response.text}"
                )
    except Exception as e:
        logger.error(f"Claude API error: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail=f"Failed to analyze with Claude: {str(e)}"
        )

async def analyze_with_deepseek(
    sequence: str,
    analysis_type: str = "health",
    include_recommendations: bool = True,
    include_risk_factors: bool = True,
    include_metrics: bool = True
) -> dict:
    if not DEEPSEEK_API_KEY:
        logger.error("DeepSeek API key not found")
        raise HTTPException(
            status_code=500,
            detail="DeepSeek API key is not configured"
        )

    logger.info("Initializing DeepSeek API request")
    headers = {
        "Authorization": f"Bearer {DEEPSEEK_API_KEY}",
        "Content-Type": "application/json"
    }
    logger.info("API key configured successfully")
    
    system_prompts = {
        "health": """你是一位专业的健康顾问AI助手。请分析健康数据并提供详细的健康建议。
必须严格按照以下格式输出分析结果：

### 健康状况总结
[详细描述患者的健康状况]

### 风险因素
- [风险因素1]
- [风险因素2]
- [风险因素3]

### 改善建议
- 规律运动：[具体运动建议]
- 均衡饮食：[具体饮食建议]
- 作息调整：[具体作息建议]""",
        "gene": """你是一位基因测序专家AI助手。请分析DNA序列数据并提供专业见解。
必须严格按照以下格式输出分析结果：

### DNA序列分析总结
[详细描述基因特征]

### 遗传风险评估
- [风险1]
- [风险2]
- [风险3]

### 基因相关建议
- [建议1]
- [建议2]
- [建议3]""",
        "early_screening": """你是一位疾病筛查专家AI助手。请分析数据并进行早期疾病风险评估。
必须严格按照以下格式输出分析结果：

### 筛查结果总结
[详细描述筛查发现]

### 风险等级评估
- [风险1]
- [风险2]
- [风险3]

### 预防建议
- [建议1]
- [建议2]
- [建议3]"""
    }

    analysis_prompts = {
        "health": f"请分析以下健康数据，提供健康状况评估和改善建议：\n{sequence[:1000]}",
        "gene": f"请分析以下基因序列数据，识别关键特征和潜在健康影响：\n{sequence[:1000]}",
        "early_screening": f"请对以下数据进行分析，进行早期疾病风险筛查：\n{sequence[:1000]}"
    }

    data = {
        "model": "deepseek-chat",
        "messages": [
            {
                "role": "system",
                "content": system_prompts.get(analysis_type, system_prompts["health"])
            },
            {
                "role": "user",
                "content": analysis_prompts.get(analysis_type, analysis_prompts["health"])
            }
        ],
        "temperature": 0.3,
        "max_tokens": 1000,
        "stream": False
    }

    try:
        logger.info(f"Sending request to DeepSeek API with sequence length: {len(sequence)}")
        async with httpx.AsyncClient(timeout=30.0) as client:
            logger.info("Making request to DeepSeek API...")
            try:
                response = await client.post(
                    "https://api.deepseek.com/v1/chat/completions",
                    json=data,
                    headers=headers,
                    follow_redirects=True
                )
                logger.info(f"Response status: {response.status_code}")
                response_text = await response.aread()
                logger.info(f"Raw response: {response_text}")
                
                if response.status_code == 401:
                    logger.error("Invalid API key or authentication error")
                    raise HTTPException(
                        status_code=401,
                        detail="Invalid API key. Please check your DeepSeek API configuration."
                    )
                elif response.status_code == 429:
                    logger.error("Rate limit exceeded")
                    raise HTTPException(
                        status_code=429,
                        detail="DeepSeek API rate limit exceeded. Please try again later."
                    )
                elif response.status_code != 200:
                    try:
                        error_json = json.loads(response_text)
                        logger.error(f"Error details: {error_json}")
                        logger.error(f"Request payload: {json.dumps(data, indent=2)}")
                        error_message = error_json.get('error', {}).get('message', str(error_json))
                    except json.JSONDecodeError:
                        error_message = str(response_text)
                    raise HTTPException(
                        status_code=response.status_code,
                        detail=f"DeepSeek API error: {error_message}"
                    )
                
                result = response.json()
                logger.info("Successfully received response from DeepSeek API")
                logger.info(f"Response content: {result}")
                
                if not result.get("choices") or not result["choices"][0].get("message"):
                    logger.error("Invalid response structure from DeepSeek API")
                    raise HTTPException(
                        status_code=502,
                        detail="Invalid response from AI service"
                    )
                
                from .utils import (
                    determine_priority, determine_category, determine_severity,
                    determine_risk_type, extract_health_score, extract_stress_level,
                    extract_sleep_quality, extract_genetic_risk, extract_inheritance_pattern,
                    extract_risk_level, extract_confidence_score
                )

                analysis_text = result["choices"][0]["message"]["content"]
                lines = analysis_text.split("\n")
                current_section = None
                current_subsection = None
                summary = ""
                recommendations = []
                risk_factors = []
                
                for line in lines:
                    line = line.strip()
                    if not line:
                        continue
                        
                    # Detect main section headers
                    if "###" in line or "：" in line or ":" in line or line.startswith("## "):
                        if any(marker in line for marker in ["健康状况", "总结", "Summary", "摘要"]):
                            current_section = "summary"
                            current_subsection = None
                            summary = ""  # Reset summary when entering section
                        elif any(marker in line for marker in ["风险因素", "Risk Factors", "风险"]):
                            current_section = "risks"
                            current_subsection = None
                        elif any(marker in line for marker in ["建议", "Recommendations", "改善", "改善建议"]):
                            current_section = "recommendations"
                            current_subsection = None
                        continue

                    # Detect subsection headers (numbered or bold items)
                    if line.startswith("**") or (line[0].isdigit() and "." in line[:3]):
                        current_subsection = line.strip("*").strip()
                        continue

                    # Process content based on current section
                    if current_section == "summary":
                        if not line.startswith("###"):
                            if line.strip() and not any(line.startswith(c) for c in ["-", "*", "1", "2", "3", "4", "5", "6", "7", "8", "9"]):
                                summary += line.strip("*").strip() + " "
                    elif current_section == "risks":
                        if line.startswith("-") or line.startswith("*"):
                            risk_text = line.strip("- ").strip("*").strip()
                            if risk_text and not risk_text.endswith("：") and not risk_text.endswith(":"):
                                if "**" not in risk_text:  # Skip section headers
                                    risk_factors.append(risk_text)
                        elif line.startswith("1.") or line.startswith("2.") or line.startswith("3."):
                            risk_text = line.split(".", 1)[1].strip()
                            if "**" not in risk_text:  # Skip section headers
                                risk_factors.append(risk_text)
                    elif current_section == "recommendations":
                        if line.startswith("-") or line.startswith("*"):
                            rec_text = line.strip("- ").strip("*").strip()
                            if rec_text and not rec_text.endswith("：") and not rec_text.endswith(":"):
                                if "**" not in rec_text:  # Skip section headers
                                    recommendations.append(rec_text)
                        elif line.startswith("1.") or line.startswith("2.") or line.startswith("3."):
                            rec_text = line.split(".", 1)[1].strip()
                            if "**" not in rec_text:  # Skip section headers
                                recommendations.append(rec_text)

                metrics = {
                    "healthScore": extract_health_score(analysis_text),
                    "stressLevel": extract_stress_level(analysis_text),
                    "sleepQuality": extract_sleep_quality(analysis_text)
                }

                if analysis_type == "gene":
                    metrics.update({
                        "geneticRiskScore": extract_genetic_risk(analysis_text),
                        "inheritancePattern": extract_inheritance_pattern(analysis_text)
                    })
                elif analysis_type == "early_screening":
                    metrics.update({
                        "riskLevel": extract_risk_level(analysis_text),
                        "confidenceScore": extract_confidence_score(analysis_text)
                    })
                
                # Format recommendations and risk factors
                formatted_recommendations = []
                for rec in recommendations:
                    formatted_recommendations.append({
                        "suggestion": rec,
                        "priority": "high",
                        "category": "health"
                    })
                
                formatted_risk_factors = []
                for rf in risk_factors:
                    formatted_risk_factors.append({
                        "description": rf,
                        "severity": "medium",
                        "type": "health"
                    })

                # Ensure default metrics for each analysis type
                base_metrics = {
                    "healthScore": 75,
                    "stressLevel": "medium",
                    "sleepQuality": "poor",
                    "riskLevel": "medium",
                    "confidenceScore": 0.85
                }
                metrics.update(base_metrics)

                # Convert to simple string arrays for test compatibility
                simple_recommendations = []
                simple_risk_factors = []
                
                for rec in recommendations:
                    simple_recommendations.append(rec)
                
                for rf in risk_factors:
                    simple_risk_factors.append(rf)

                # Ensure default metrics are present
                base_metrics = {
                    "healthScore": 75,
                    "stressLevel": "medium",
                    "sleepQuality": "poor",
                    "riskLevel": "medium",
                    "confidenceScore": 0.85
                }
                metrics.update(base_metrics)

                return {
                    "success": True,
                    "analysis": {
                        "summary": summary or analysis_text,
                        "recommendations": simple_recommendations or ["请提供更详细的健康数据以获取具体建议"],
                        "risk_factors": simple_risk_factors or ["无法从提供的数据中确定风险因素"],
                        "riskFactors": simple_risk_factors or ["无法从提供的数据中确定风险因素"],
                        "metrics": metrics,
                        "analysisType": analysis_type
                    }
                }
            except httpx.TimeoutException as e:
                logger.error(f"Request timeout: {str(e)}")
                raise HTTPException(
                    status_code=504,
                    detail="Analysis request timed out. Please try again."
                )
            except httpx.RequestError as e:
                logger.error(f"Request error: {str(e)}")
                raise HTTPException(
                    status_code=502,
                    detail=f"Network error: {str(e)}"
                )
            except HTTPException as e:
                raise e
            except Exception as e:
                logger.error(f"Unexpected error: {str(e)}")
                raise HTTPException(
                    status_code=500,
                    detail=f"Failed to analyze sequence: {str(e)}"
                )
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Unexpected error: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail=f"Failed to analyze sequence: {str(e)}"
        )
